src/goto-programs/remove_function_pointers.cpp

```diff
@@ -140,6 +162,341 @@ symbolt &remove_function_pointerst::new_tmp_symbol()
 
 /*******************************************************************\
 
+Function: remove_function_pointerst::try_get_precise_call
+
+  Inputs:
+   expr - The expression that is the value of the function pointer.
+
+ Outputs: Returns true if the expression is a symbol and its value was code.
+          In this case out_function will be the function this is a symbol for.
+
+ Purpose: Find if a symbol is in fact a function that can be used in place
+          of the function pointer
+
+\*******************************************************************/
+
+bool remove_function_pointerst::try_get_precise_call(
+  const exprt &expr,
+  exprt &out_function)
+{
+  if(expr.id()==ID_symbol && expr.type().id()==ID_code)
+  {
+    out_function=to_symbol_expr(expr);
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_function_pointerst::try_get_from_address_of
+
+  Inputs:
+   expr - The expression that is the value of the function pointer.
+
+ Outputs: Returns true if the expression is a address_of_exprt and it was able
+          to either pointing to a function or a sybmol. In either of
+          the true cases the out_functions will contain all the functions the
+          address of could be.
+
+ Purpose: Resolve a address of to its real value.
+
+\*******************************************************************/
+
+bool remove_function_pointerst::try_get_from_address_of(
+  const exprt &expr, functionst &out_functions)
+{
+  if(expr.id()==ID_address_of)
+  {
+    const address_of_exprt &address_of_expr=to_address_of_expr(expr);
+
+    exprt precise_expr;
+    if(try_get_precise_call(address_of_expr.object(), precise_expr))
+    {
+      out_functions.push_back(precise_expr);
+      return true;
+    }
+    else if(try_get_call_from_symbol(address_of_expr.object(), out_functions))
+    {
+      return true;
+    }
+    else
```

Comment:
The following lines could be replaced by a `return false` on the last line of the function, do people think it is clearer like this or collapsed down. 
Clearer like it is - can see all the ways the method can fail explicitly
Clearer other way - hard to keep track of which if the else corresponds to.